wolfcrypt/src/port/atmel/atmel.c

```diff
@@ -931,72 +942,131 @@ int atcatls_verify_signature_cb(WOLFSSL* ssl, const byte* sig, unsigned int sigS
+       printf("Failed to get max signer cert size\r\n");
+    #endif
+       return ret;
     }
-
-    /* fetch device cert size */
-    status=tng_atcacert_read_device_cert(NULL, &deviceCertSize, NULL);
-    if (ATCA_SUCCESS != status) {
-        #ifdef WOLFSSL_ATECC_DEBUG
-        printf("Failed reading device cert size(0x%x)\r\n", status);
-        #endif
-        return WOLFSSL_FAILURE;
-    }    
-    certBuffer=(uint8_t*)XMALLOC(signerCertSize+deviceCertSize, NULL, DYNAMIC_TYPE_TMP_BUFFER);
-    if(NULL == certBuffer){
-        #ifdef WOLFSSL_ATECC_DEBUG
-        printf("Failed allocating space for certBuffer\r\n");
-        #endif
-        return WOLFSSL_FAILURE;
+    else if (signerCertSize > ATCATLS_SIGNER_CERT_MAX_SIZE) {
+    #ifdef WOLFSSL_ATECC_DEBUG
+        printf("Signer CA cert buffer too small, need to increase at least"
+               " to %d\r\n", signerCertSize);
+    #endif
+       return -1;
     }
-    
-    /* Read signer cert */
-    status = tng_atcacert_read_signer_cert(&certBuffer[deviceCertSize],
-                                           &signerCertSize);
+
+    /* Read TNGTLS signer cert */
+    status = tng_atcacert_read_signer_cert(signerBuffer, &signerCertSize);
     if (ATCA_SUCCESS != status) {
-        #ifdef WOLFSSL_ATECC_DEBUG
-        printf("Error reading signer cert(0x%x)\r\n", status);
-        #endif
-        XFREE(certBuffer,NULL,DYNAMIC_TYPE_TMP_BUFFER);
         ret = atmel_ecc_translate_err(ret);
         return ret;
     }
 
     /* Read device cert signed by the signer above */
     status = tng_atcacert_read_device_cert(certBuffer, &deviceCertSize,
-     	                                   &certBuffer[deviceCertSize]);
+                                           signerBuffer);
     if (ATCA_SUCCESS != status) {
-        #ifdef WOLFSSL_ATECC_DEBUG
-        printf("Error reading device cert(0x%x)\r\n", status);
-        #endif
-        XFREE(certBuffer,NULL,DYNAMIC_TYPE_TMP_BUFFER);
         ret = atmel_ecc_translate_err(ret);
         return ret;
     }
+    else if (deviceCertSize > ATCATLS_DEVICE_CERT_MAX_SIZE) {
+    #ifdef WOLFSSL_ATECC_DEBUG
+        printf("Device cert buffer too small, need to increase at least"
+               " to %d\r\n", deviceCertSize);
+    #endif
+       return -1;
+    }
+#endif
+
+#ifdef WOLFSSL_ATECC_TFLXTLS
+    /* MAKE SURE TO COPY YOUR CUSTOM CERTIFICATE FILES UNDER CAL/tng
+     * Verify variable names, here below the code uses typical tflxtls
+     *  proto example.
+     *
+     * g_cert_def_1_signer
+     * g_cert_ca_public_key_1_signer
+     * g_cert_def_3_device
+     */
+
+    status = atcacert_read_cert(&g_cert_def_1_signer,
+                            (const uint8_t*)g_cert_ca_public_key_1_signer,
+                            signerBuffer, &signerCertSize);
+    if (status != ATCA_SUCCESS) {
+    #ifdef WOLFSSL_ATECC_DEBUG
+        printf("Failed to read TFLXTLS signer cert!\r\n");
+    #endif
+        return ret = status;
```

Comment:
This isn't typical coding convention, but it works. Why not just `return (int)status;`?